Make it possible to extend the size of the heap

rafalmiel opened the pull request:
Add a max_size parameter to the heap.
Allocations exceeding the size of the heap will fail,
but one can extend the size of the heap with the call to extend.

That allows heap to be more memory efficient, consuming more memory only
when necessary, eg. os allocator might map additional virtual memory when
the allocation fails.

oli-obk commented:
How is this helpful without some callback asking whether the memory can be extended? I mean the memory can't be used according to the docs, so why not simply add it to the heap from the beginning?

rafalmiel commented:
The intention was that the extend would be called by the user of the heap after allocating additional memory for the new extended heap. I would like to have initial small (bottom, size) area and when the allocation fails I allocate additional virtual address space at the end. Otherwise I would have to provide memory for (bottom, max_size) from the beginning, which might be wasteful if there will not be many allocations. Think of it as a sbrk call on linux

oli-obk commented:
Ah, yea that makes sense. I got the API backwards in my head.

phil-opp commented on src/hole.rs:
```diff
@@ -40,6 +40,38 @@ impl HoleList {
+
+    /// Extends the size of the list by increasing the size of the last hole
+    /// or appending new hole at the end of the list
+    pub fn extend(&mut self, addr: usize, size: usize) {
```
Why do we need `extend` and `put_hole`? Can't we just use the `deallocate` function for this?

rafalmiel replied:
Deallocate should work indeed, good catch. Perhaps it's just not a good name for this functionality? How about extend calling deallocate, or should we just get rid of extend call altogether?

phil-opp replied:
> How about extend calling deallocate, or should we just get rid of extend call altogether?

I'm fine with both variants.

phil-opp commented on src/lib.rs:
```diff
@@ -17,6 +17,7 @@ mod test;
 pub struct Heap {
     bottom: usize,
     size: usize,
+    max_size: usize,
```
For what do we need `max_size`? The only reason I can see is that this makes it possible to make `extend` safe. Maybe it's better to have an `unsafe` extend function and let the caller manage `max_size` itself.

(Maybe we should even get rid of the `bottom` and `size` fields… Then we could expand the heap with an arbitrary memory chunk without the need for a single continuous block.)

rafalmiel replied:
Thanks for the review! Yes, that makes sense, we could leave it to the user to manage the complexity of the heap size. Also, one can then choose between allocating small heap size or going for on demand mapping. I will post the revised version in the coming days. Cheers

phil-opp replied:
Thanks, looking forward to it! :)

rafalmiel commented on src/hole.rs:
```diff
@@ -40,6 +40,38 @@ impl HoleList {
 
+    /// Places a hole of a given size at a given address after a given hole
+    unsafe fn put_hole(hole: &mut Hole, addr: usize, size: usize) {
+        mem::forget(mem::replace(&mut *(addr as *mut Hole),
```
Not sure if it is really needed here, but isn't it the simillar case as in the new call? It seems to work without the forget call though(because Hole does not have a drop Trait?). Still exploring the mysteries of rust :)

phil-opp replied:
> because Hole does not have a drop Trait?

Yes, exactly. I'm not sure why I used it in `new`… maybe `Hole` had a destructor back then and I forgot to remove it later. In any case, it shouldn't be needed here.

rafalmiel commented:
Hello, 

I updated the pull request. As suggested I removed the max_size field and made the call to the extend unsafe. Also I make use of the deallocate method now (fixed some issue there. If the new hole was added just behind the last hole, the last hole should be merged, instead new hole was being created). I hope it looks better now :)

I have also separate branch where I removed bottom and size fields completely, but feel like it's a big and potentially breaking change for client code.

You can have a look into it here
https://github.com/rafalmiel/linked-list-allocator/commit/28764f4f6ac3b9d2e3ddf229eda3b0a53c83121d
if you wish I will also include it in the pull request.

Thanks!

phil-opp commented:
Thanks a lot for the updates! Looks really good now.

I'm a bit short of time at the moment, but I try to take a closer look in the next few days.

rafalmiel commented:
Thanks a lot, no need to hurry :)

phil-opp commented on src/hole.rs:
```diff
@@ -271,17 +271,21 @@ fn deallocate(mut hole: &mut Hole, addr: usize, mut size: usize) {
-                hole.next = Some(unsafe { Unique::new(ptr) });
+                if hole_addr + hole.size == addr {
+                    // after:   ___XXXFFFF___    where F is the freed block
+                    hole.size += size;
```
Good catch! I think we could merge this case with [lines 243-249](https://github.com/rafalmiel/linked-list-allocator/blob/b7dc99795964afed5a2ab64f6bc8d6c6e193e89e/src/hole.rs#L243-L249) by replacing the `Some(_)` with a `None`. Or am I missing something?

rafalmiel commented:
> Good catch! I think we could merge this case with lines 243-249 by replacing the Some(_) with a None. Or am I missing something?

Hi Phil, 

I gave it a second thought and updated the pull request, but instead of None I went with _. I think if we do None, then we will miss the case when the new hole aligns to previous hole, but there are more holes later on the heap?

I hope it looks good, please let me know if I can improve it more.

Thanks,
Rafal

phil-opp commented on src/hole.rs:
```diff
@@ -264,6 +257,17 @@ fn deallocate(mut hole: &mut Hole, addr: usize, mut size: usize) {
                 hole = move_helper(hole).next_unwrap(); // start next iteration at next hole
                 continue;
             }
+            _ if hole_addr + hole.size == addr => {
```
Is there a reason why you moved this match case to the bottom? I think it should work too, but I'm not 100% sure… 

(The order matters for some cases, e.g. the `XXXFFFFYYYYY` case must come before the `XXXFFFF__YYYYY` and `XXX__FFFFYYYYY` cases.)

rafalmiel replied:
No big reason, just wanted to keep it close to the other _ case, but I am happy to move it back
Thanks!

phil-opp replied:
Ok, then I would move it back

rafalmiel commented:
Hi, I updated the pull request with the requested change to the match statement

phil-opp commented:
Thanks a lot!

phil-opp commented:
Published as version 0.2.7
